File: 3.Serialization-and-Deserialization-or-backend-engineers/code/typescript/serialization_in_go_1.ts
// ABSTRACTION: Serializer names a capability - "turn data into
// bytes and back" - without binding to a concrete format. Code
// depends on this contract, so JSON could be swapped for another
// format with zero changes to callers.
export interface Serializer {
    serialize(v: any): string;         // native -> common format
    deserialize(data: string): any;    // common format -> native
}

// POLYMORPHISM: JSONSerializer implements Serializer. Any other
// format (YAML, Protobuf) could implement the same interface and
// be used interchangeably through a Serializer variable.
export class JSONSerializer implements Serializer {
    public serialize(v: any): string {
        return JSON.stringify(v); // SERIALIZE: object -> JSON string
    }
    public deserialize(data: string): any {
        return JSON.parse(data);  // DESERIALIZE: JSON string -> object
    }
}

// "INHERITANCE" / COMPOSITION: BaseModel holds shared fields;
// extending it reuses them.
export class BaseModel {
    public id: number = 0;
    public created_at: Date = new Date();
}

export interface Address {
    country: string;
    phone: number;
}

// ENCAPSULATION: fields can be private (or excluded via toJSON).
// We use a custom toJSON method to ensure the password never leaks
// over the wire.
export class User extends BaseModel {
    public name: string;
    public active: boolean;
    public address: Address;
    private password_: string; // hidden from JSON output

    constructor(id: number, name: string, active: boolean, address: Address, password_: string) {
        super();
        this.id = id;
        this.name = name;
        this.active = active;
        this.address = address;
        this.password_ = password_;
    }

    // Control the JSON shape
    public toJSON(): any {
        return {
            id: this.id,
            created_at: this.created_at.toISOString(),
            name: this.name,
            active: this.active,
            address: this.address
        };
    }
}

if (require.main === module) {
    const codec: Serializer = new JSONSerializer(); // program to the interface

    const u = new User(
        1,
        "Ada",
        true,
        { country: "India", phone: 123456 },
        "supersecret"
    );

    // SERIALIZE - native object into the common JSON format
    const out = codec.serialize(u);
    console.log(out);
    // {"id":1,"created_at":"...","name":"Ada","active":true,
    //  "address":{"country":"India","phone":123456}}

    // DESERIALIZE - JSON received over HTTP back into a native object
    const incoming = '{"name":"Lin","address":{"country":"IN","phone":42}}';
    const back = codec.deserialize(incoming);
    console.log(back.name, back.address.country); // Lin IN
}

File: 3.Serialization-and-Deserialization-or-backend-engineers/code/rust/serialization_in_go_1.rs
use serde::{Deserialize, Serialize};
use serde_json;
use chrono::{DateTime, Utc};

// ABSTRACTION: Serializer names a capability - "turn data into
// bytes and back" - without binding to a concrete format. Code
// depends on this contract, so JSON could be swapped for another
// format with zero changes to callers.
pub trait Serializer {
    fn serialize<T: Serialize>(&self, v: &T) -> Result<String, String>;
    fn deserialize<'a, T: Deserialize<'a>>(&self, data: &'a str) -> Result<T, String>;
}

// POLYMORPHISM: JSONSerializer implements Serializer. Any other
// format (YAML, Protobuf) could implement the same interface and
// be used interchangeably through a Serializer variable.
pub struct JSONSerializer;

impl Serializer for JSONSerializer {
    fn serialize<T: Serialize>(&self, v: &T) -> Result<String, String> {
        serde_json::to_string(v).map_err(|e| e.to_string()) // SERIALIZE: struct -> JSON string
    }
    
    fn deserialize<'a, T: Deserialize<'a>>(&self, data: &'a str) -> Result<T, String> {
        serde_json::from_str(data).map_err(|e| e.to_string()) // DESERIALIZE: JSON string -> struct
    }
}

// "INHERITANCE" via COMPOSITION: BaseModel holds shared fields;
// embedding it into User reuses them (and their json tags).
#[derive(Serialize, Deserialize, Debug)]
pub struct BaseModel {
    #[serde(default)]
    pub id: i32,
    #[serde(default = "chrono::Utc::now")]
    pub created_at: DateTime<Utc>,
}

// ENCAPSULATION: an exported struct whose JSON shape is controlled
// by macros. `password` uses `skip` -> private AND
// invisible to the JSON encoder, so it never leaks over the wire.
#[derive(Serialize, Deserialize, Debug)]
pub struct User {
    #[serde(flatten)]
    pub base: BaseModel,            // embedded -> inherits ID, CreatedAt
    pub name: String,
    
    #[serde(default = "default_active")]
    pub active: bool,
    
    pub address: Address,           // nested object
    
    #[serde(skip, default)]
    password: String,               // unexported: hidden from JSON output
}

fn default_active() -> bool { true }

#[derive(Serialize, Deserialize, Debug)]
pub struct Address {
    pub country: String,
    pub phone: i32,
}

fn main() {
    let codec = JSONSerializer; // program to the interface

    let u = User {
        base: BaseModel { id: 1, created_at: Utc::now() },
        name: "Ada".to_string(),
        active: true,
        address: Address { country: "India".to_string(), phone: 123456 },
        password: "supersecret".to_string(),
    };

    // SERIALIZE - native Rust struct into the common JSON format
    let out = codec.serialize(&u).unwrap();
    println!("{}", out);
    // {"id":1,"created_at":"...","name":"Ada","active":true,
    //  "address":{"country":"India","phone":123456}}

    // DESERIALIZE - JSON received over HTTP back into a Rust struct
    let incoming = r#"{"name":"Lin","address":{"country":"IN","phone":42}}"#;
    let back: User = codec.deserialize(incoming).unwrap();
    println!("{} {}", back.name, back.address.country); // Lin IN
}
